chainer/iterators/serial_iterator.py

```diff
@@ -107,22 +133,25 @@ def serialize(self, serializer):
         except KeyError:
             # guess previous_epoch_detail for older version
             self._previous_epoch_detail = self.epoch + \
-                (self.current_position - self.batch_size) / len(self.dataset)
+                (self.current_position - self.batch_size) / self._epoch_size
             if self.epoch_detail > 0:
                 self._previous_epoch_detail = max(
                     self._previous_epoch_detail, 0.)
             else:
                 self._previous_epoch_detail = -1.
 
     def reset(self):
-        if self._shuffle:
-            self._order = numpy.random.permutation(len(self.dataset))
-        else:
-            self._order = None
-
         self.current_position = 0
         self.epoch = 0
         self.is_new_epoch = False
 
         # use -1 instead of None internally.
         self._previous_epoch_detail = -1.
+        self._order = self.order_sampler(numpy.arange(len(self.dataset)), 0)
```

Comment:
I'm starting to wonder if we can get rid of the `NoShuffleOrderSampler` class completely if we change this logic to e.g. `self._order = self.order_sampler(...) if self.order_sampler is not None else None`. Would that be possible you think?

Reply:
Yes. This simplifies the interface of OrderSampler